scm_v3c/applications/ble_tx_154_rx_track_if/ble_tx_154_rx_track_if.c

```diff
@@ -0,0 +1,300 @@
+    }
+}
+
+//=========================== public ==========================================
+
+//=========================== private =========================================
+
+void    cb_startFrame_rx(uint32_t timestamp){
+    app_vars.rxFrameStarted = true;
+}
+
+void    cb_endFrame_rx(uint32_t timestamp){
+
+    uint8_t i;
+    uint32_t IF_estimate, LQI_chip_errors;
+
+    radio_getReceivedFrame(
+        &(app_vars.packet[0]),
+        &app_vars.packet_len,
+        sizeof(app_vars.packet),
+        &app_vars.rxpk_rssi,
+        &app_vars.rxpk_lqi
+    );
+
+    // Read IF estimate and LQI chip errors.
+    IF_estimate       = radio_getIFestimate();
+    LQI_chip_errors   = radio_getLQIchipErrors();
+
+    radio_rfOff();
+
+    if (
+        app_vars.packet_len == LEN_RX_PKT && (radio_getCrcOk())
+    ) {
+        // Only record IF estimate, LQI chip errors, and CDR tau for valid packets.
+        app_vars.IF_estimate        = IF_estimate;
+        app_vars.LQI_chip_errors    = LQI_chip_errors;
+
+        printf(
+            "pkt received on ch%d %d%d%d%d.%d.%d.%d\r\n",
+            app_vars.packet[4],
+            app_vars.packet[0],
+            app_vars.packet[1],
+            app_vars.packet[2],
+            app_vars.packet[3],
+            app_vars.rx_coarse,
+            app_vars.rx_mid,
+            app_vars.rx_fine
+        );
+        printf("IF estimate: %u, LQI: %d\n", app_vars.IF_estimate, app_vars.LQI_chip_errors);
+
+        ble_set_data(app_vars.packet);
+        ble_set_data_tx_en(true);
+
+        app_vars.packet_len = 0;
+        memset(&app_vars.packet[0], 0, LENGTH_PACKET);
+
+        app_vars.rx_received = true;
+        ++app_vars.rx_num_received;
+
+        radio_update_IF_estimate(app_vars.IF_estimate, app_vars.LQI_chip_errors);
+    }
+
+    radio_rxEnable();
+    radio_rxNow();
+
+    app_vars.rxFrameStarted = false;
+}
+
+void    cb_timer(void) {
+    app_vars.changeConfig = true;
+}
+
+void    receive_154_packet(void) {
+    printf("Receiving on %u %u %u\n", app_vars.rx_coarse, app_vars.rx_mid, app_vars.rx_fine);
+    while (app_vars.rxFrameStarted);
+    radio_rfOff();
+    app_vars.rx_received = false;
+    app_vars.rx_num_received = 0;
+    LC_FREQCHANGE(app_vars.rx_coarse, app_vars.rx_mid, app_vars.rx_fine);
+    radio_rxEnable();
+    radio_rxNow();
+}
+
+void    transmit_ble_packet(void) {
+    int t;
+
+    ble_gen_packet();
+
+    LC_FREQCHANGE(app_vars.tx_coarse, app_vars.tx_mid, app_vars.tx_fine);
+    printf("Transmitting on %u %u %u\n", app_vars.tx_coarse, app_vars.tx_mid, app_vars.tx_fine);
+
+    // Wait for frequency to settle.
+    for (t = 0; t < 5000; ++t);
+
+    ble_transmit();
+}
+
+void    calibrate_fine_code(void) {
+    if (radio_get_IF_estimate_ready()) {
+        int diff;
+
+        app_vars.IF_offset_prev = app_vars.IF_offset;
+        app_vars.IF_offset = radio_get_IF_estimate();
+
+        diff = app_vars.IF_offset / 12; // empirically tuned
+        app_vars.rx_fine += diff;
```

Comment:
This only assumes the LC oscillator drifts within a range that only need to tune the  fine code, right? When it is not, such as a large temperature change, mid code calibration may need.

Reply:
Yes, but since the LC frequency code is not monotonic, I think in order to support a large temperature range, it's probably best to find the coarse/mid codes for different ranges of temperature and then do this "local" fine code calibration with the previously found coarse/mid code in this temperature range.